File: server/app/api/v1/practice_sessions.py
# server/app/api/v1/practice_sessions.py
# The practice-session lifecycle surface — FLE-21 §5's call list.
#
# This is the HTTP boundary for pilot telemetry. FLE-13 names day-7 return and
# session-completion rate as the numbers that count, and FLE-10 forbids an analytics
# SDK in a pilot build, so every one of those numbers is derived from a row one of
# these five handlers wrote. If a handler here silently drops a write, the readout
# degrades to self-report and nobody finds out until it is being written.
#
# NOT the Phase-3 ratings endpoint. `POST /api/v1/sessions` in sessions.py is the
# Today-card rating and is untouched by all of this (FLE-21 §5.2) — it keeps serving
# ratings taken OUTSIDE the player. Do not extend it and do not route to it from here.
#
# Shared file with FLE-63. That issue owns the two generate/read endpoints —
# `POST /practice-sessions/today` (a thin wrapper over store.resolve_or_generate) and
# `GET /practice-sessions/{id}` — and they belong in this router rather than a second
# one. ⚠️ When `GET /practice-sessions/{session_id}` is added it MUST be declared
# BELOW `GET /practice-sessions/current`: FastAPI matches in declaration order, and a
# `{session_id}` path registered first would swallow `/current` and fail it as an
# invalid UUID.
#
# Failure semantics are SPLIT, and this is the part that is easy to flatten by
# accident (FLE-21 §5.2):
#
#   /enter, /skip, /complete with rating=null → fire-and-forget. The player queues and
#       retries; a permanent failure is one lost telemetry row, which is acceptable.
#   /complete with a non-null rating          → a USER-VISIBLE write. The player
#       surfaces the outcome, and 409 keeps UI-SPEC §10's existing meaning: a
#       state-sync signal, not an error toast.
#
# Which is why every handler here is idempotent and order-tolerant rather than
# strict. See lifecycle.py's header for the lattice that makes that true.
import logging
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_tz_offset_minutes, get_user_id
from app.db.session import get_db
from app.models.practice_session import (
    ItemCompleteRequest,
    ItemEventResponse,
    ItemSkipRequest,
    PracticeSessionItemResponse,
    PracticeSessionResponse,
    SessionCompleteResponse,
)
from app.sessions import lifecycle
from app.sessions.lifecycle import (
    ItemNotFound,
    RatingNotPermitted,
    SessionNotFound,
    SessionRow,
)
from app.sessions.snapshot import local_day

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/practice-sessions", tags=["practice-sessions"])


def _iso(value) -> str | None:
    return None if value is None else value.isoformat()


def _session_payload(session: SessionRow, items: list) -> PracticeSessionResponse:
    return PracticeSessionResponse(
        id=session.id,
        user_id=session.user_id,
        local_calendar_day=session.local_calendar_day.isoformat(),
        tz_offset_minutes=session.tz_offset_minutes,
        song_id=session.song_id,
        target_minutes=session.target_minutes,
        mode=session.mode,
        state=session.state,
        terminal_reason=session.terminal_reason,
        generated_at=_iso(session.generated_at),
        started_at=_iso(session.started_at),
        last_activity_at=_iso(session.last_activity_at),
        ended_at=_iso(session.ended_at),
        completion_ratio=(
            None if session.completion_ratio is None else float(session.completion_ratio)
        ),
        elapsed_active_seconds=session.elapsed_active_seconds,
        item_count=session.item_count,
        last_item_index_reached=session.last_item_index_reached,
        items=[
            PracticeSessionItemResponse(
                item_index=i.item_index,
                block=i.block,
                kind=i.kind,
                state=i.state,
                drill_id=i.drill_id,
                song_id=i.song_id,
                target_skill_node_id=i.target_skill_node_id,
                planned_seconds=i.planned_seconds,
                planned_bpm=i.planned_bpm,
                planned_reps=i.planned_reps,
                completed_reps=i.completed_reps,
                rated=i.rated,
                skippable=i.skippable,
                click_enabled=i.click_enabled,
                re_entry=i.re_entry,
                repeat_ok=i.repeat_ok,
                is_consolidation=i.is_consolidation,
                started_at=_iso(i.started_at),
                ended_at=_iso(i.ended_at),
                active_seconds=i.active_seconds,
                rating=i.rating,
                advance_mode=i.advance_mode,
            )
            for i in items
        ],
    )


async def _payload(db: AsyncSession, session: SessionRow) -> PracticeSessionResponse:
    return _session_payload(session, await lifecycle.load_items(db, session.id))


def _not_found(exc: Exception) -> HTTPException:
    return HTTPException(status_code=404, detail=str(exc))


# ---------------------------------------------------------------------------
# Read
# ---------------------------------------------------------------------------


@router.get("/current", response_model=PracticeSessionResponse)
async def get_current(
    user_id: UUID = Depends(get_user_id),
    tz_offset_minutes: int = Depends(get_tz_offset_minutes),
    db: AsyncSession = Depends(get_db),
) -> PracticeSessionResponse:
    """The caller's open session for their local day, or 404.

    Read-only by contract (FLE-21 §5) — it NEVER writes, not even to bump
    `last_activity_at`. A resume check that mutated would make merely opening the app
    indistinguishable from practising, and `last_activity_at` is what the §3 sweep
    reads to decide whether a session was abandoned.

    404 is the normal answer on a day the user has not generated a session yet; it is
    not an error condition for the player.
    """
    day = await local_day(db, tz_offset_minutes)
    session = await lifecycle.get_current_session(db, user_id, today=day)
    if session is None:
        raise HTTPException(
            status_code=404, detail="No open practice session for today."
        )
    return await _payload(db, session)


# ---------------------------------------------------------------------------
# Lifecycle writes
# ---------------------------------------------------------------------------


@router.post("/{session_id}/start", response_model=PracticeSessionResponse)
async def start(
    session_id: UUID,
    user_id: UUID = Depends(get_user_id),
    db: AsyncSession = Depends(get_db),
) -> PracticeSessionResponse:
    """Mark the session opened. Write-once, idempotent (FLE-21 R5).

    `started_at` is the day-7 return metric, so a second call — a retry, or a resume
    after the user put the phone down — must NOT move it. Answers 200 either way; a
    no-op is a success, not a conflict.
    """
    try:
        async with db.begin():
            session = await lifecycle.start_session(db, user_id, session_id)
    except SessionNotFound as exc:
        raise _not_found(exc) from exc
    return await _payload(db, session)


@router.post("/{session_id}/items/{item_index}/enter", response_model=ItemEventResponse)
async def enter_item(
    session_id: UUID,
    item_index: int,
    user_id: UUID = Depends(get_user_id),
    db: AsyncSession = Depends(get_db),
) -> ItemEventResponse:
    """Item → in_progress; bump the session's high-water mark.

    A late `enter` arriving behind its own `complete` (the outbox can flush out of
    order) is a no-op 200 with `applied=false` — it must not resurrect a finished item.
    """
    try:
        async with db.begin():
            outcome, session = await lifecycle.enter_item(
                db, user_id, session_id, item_index
            )
    except SessionNotFound as exc:
        raise _not_found(exc) from exc
    except ItemNotFound as exc:
        raise _not_found(exc) from exc
    return ItemEventResponse(
        item_index=outcome.item_index,
        state=outcome.state,
        applied=outcome.applied,
        active_seconds=outcome.active_seconds,
        clamped=outcome.clamped,
        session=await _payload(db, session),
    )


@router.post(
    "/{session_id}/items/{item_index}/complete", response_model=ItemEventResponse
)
async def complete_item(
    body: ItemCompleteRequest,
    session_id: UUID,
    item_index: int,
    user_id: UUID = Depends(get_user_id),
    db: AsyncSession = Depends(get_db),
) -> ItemEventResponse:
    """Item → completed, with the rating if there is one (FLE-21 §5.1).

    `rating: null` is valid and still lands `completed` (R1) — warm-up, `song_play`,
    consolidation and any slot dropped by MAX_RATING_TAPS are unrated BY DESIGN, and
    in a 15-minute session that is most of the items.

    422 `rating_not_permitted_for_item` when a rating arrives for an item whose
    server-side `rated` flag is false. That can only come from a client bug, and the
    player's outbox must treat it as terminal — drop it, do not retry.
    """
    try:
        async with db.begin():
            outcome, session = await lifecycle.complete_item(
                db,
                user_id,
                session_id,
                item_index,
                active_seconds=body.active_seconds,
                rating=body.rating,
                completed_reps=body.completed_reps,
                advance_mode=body.advance_mode,
            )
    except SessionNotFound as exc:
        raise _not_found(exc) from exc
    except ItemNotFound as exc:
        raise _not_found(exc) from exc
    except RatingNotPermitted as exc:
        raise HTTPException(
            status_code=422, detail=f"rating_not_permitted_for_item: {exc}"
        ) from exc
    return ItemEventResponse(
        item_index=outcome.item_index,
        state=outcome.state,
        applied=outcome.applied,
        active_seconds=outcome.active_seconds,
        clamped=outcome.clamped,
        session=await _payload(db, session),
    )


@router.post("/{session_id}/items/{item_index}/skip", response_model=ItemEventResponse)
async def skip_item(
    body: ItemSkipRequest,
    session_id: UUID,
    item_index: int,
    user_id: UUID = Depends(get_user_id),
    db: AsyncSession = Depends(get_db),
) -> ItemEventResponse:
    """Item → skipped. Its own call, never inferred from navigation (FLE-21 §5.4).

    This is the endpoint FLE-13's "which block do they skip" is made of. Advancing
    past an item is not a skip unless the user chose to pass — an auto-advance goes to
    `/complete` with `advance_mode='auto'` instead.
    """
    try:
        async with db.begin():
            outcome, session = await lifecycle.skip_item(
                db,
                user_id,
                session_id,
                item_index,
                active_seconds=body.active_seconds,
                completed_reps=body.completed_reps,
            )
    except SessionNotFound as exc:
        raise _not_found(exc) from exc
    except ItemNotFound as exc:
        raise _not_found(exc) from exc
    return ItemEventResponse(
        item_index=outcome.item_index,
        state=outcome.state,
        applied=outcome.applied,
        active_seconds=outcome.active_seconds,
        clamped=outcome.clamped,
        session=await _payload(db, session),
    )


@router.post("/{session_id}/complete", response_model=SessionCompleteResponse)
async def complete_session(
    session_id: UUID,
    user_id: UUID = Depends(get_user_id),
    db: AsyncSession = Depends(get_db),
) -> SessionCompleteResponse:
    """Session → completed, with the summary numbers (FLE-10 R8).

    Outranks the clock (FLE-21 §3.1): a user who practised across midnight gets
    day-rolled underneath them by the sweep, and when they then tap "done" that is the
    truth. Since R7 deleted `superseded`, the clock is the only thing that can abandon
    a session — so every abandonment is overridable here, and only here.
    """
    try:
        async with db.begin():
            session, tally = await lifecycle.complete_session(db, user_id, session_id)
    except SessionNotFound as exc:
        raise _not_found(exc) from exc
    return SessionCompleteResponse(
        session=await _payload(db, session),
        completion_ratio=(
            None if session.completion_ratio is None else float(session.completion_ratio)
        ),
        done_items=tally.done_items,
        skipped_items=tally.skipped_items,
        planned_items=tally.planned_items,
    )

File: mobile/src/api/practiceSessions.ts
// mobile/src/api/practiceSessions.ts
// TanStack Query hooks over /api/v1/practice-sessions/* (FLE-10 Task 6).
//
// Contract source: FLE-63 (routes + schemas) + FLE-21 (lifecycle writes), reconciled
// against the FLE-10 design doc (§10 R1-R10). Full rulings are in the docstrings on
// schema.d.ts's `operations` entries — this file does not repeat them, it implements them.
//
// FLE-10 R5 (client derives nothing): every hook here is a thin wrapper over one HTTP
// call. No hook computes state, block order, ratios, or ladder outcomes — those are
// server-told fields on the response and the caller reads them directly.
//
// Query key shape:
//   ['practice-session', sessionId]                    — one session, by id (resume-by-id, summary re-read)
//   ['practice-session', 'current', userId, localDay]   — today's open session, if any (404 is normal)
//
// Every mutation that returns a fresh PracticeSessionResponse writes it into BOTH keys
// (when a session id is known) so the walker and the "resume?" check on Today never
// disagree with each other after a transition.
//
// What this file deliberately does NOT build yet: the offline outbox (MMKV-backed FIFO
// queue with retry) that design doc §5/§14.4 calls for. These hooks are the network
// calls the outbox will eventually wrap; the walker screens land in a follow-up slice.
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import type { components } from './generated/schema';
import { apiFetch, apiFetchWithStatus } from './apiClient';
import { getOrCreateUserId } from './mmkv';
import { localCalendarDay } from './todaySong';

export type PracticeSessionResponse = components['schemas']['PracticeSessionResponse'];
export type PracticeSessionItemResponse = components['schemas']['PracticeSessionItemResponse'];
export type ItemEventResponse = components['schemas']['ItemEventResponse'];
export type SessionCompleteResponse = components['schemas']['SessionCompleteResponse'];
export type EmbeddedDrill = components['schemas']['EmbeddedDrill'];
export type EmbeddedSong = components['schemas']['EmbeddedSong'];
export type PracticeRatingLiteral = 'not_my_tempo' | 'getting_closer' | 'thats_what_im_looking_for';

export function practiceSessionQueryKey(sessionId: string) {
  return ['practice-session', sessionId] as const;
}

export function currentPracticeSessionQueryKey(userId: string, day: string) {
  return ['practice-session', 'current', userId, day] as const;
}

/**
 * First not-yet-finished item, server-authoritative (design doc §5 rule 1): "the
 * server's first non-terminal item wins — always. It decides where you are." Used
 * by both the session entry route (where to land) and the walker (where to go
 * next after a complete/skip response carries a fresh item list).
 *
 * Returns null when every item is completed/skipped — the session is done.
 */
export function firstNonTerminalIndex(
  items: PracticeSessionItemResponse[] | undefined,
): number | null {
  if (!items) return null;
  const idx = items.findIndex((it) => it.state === 'not_reached' || it.state === 'in_progress');
  return idx === -1 ? null : idx;
}

/** Writes a fresh session snapshot into both the by-id and the current-day cache slots. */
function cacheSession(
  qc: ReturnType<typeof useQueryClient>,
  userId: string,
  session: PracticeSessionResponse,
) {
  qc.setQueryData(practiceSessionQueryKey(session.id), session);
  qc.setQueryData(currentPracticeSessionQueryKey(userId, session.local_calendar_day), session);
}

/**
 * Resolve-or-generate today's session (FLE-63). The one call that starts a session.
 *
 * A mutation, not a query, because it can write (a fresh plan) — but unlike a typical
 * mutation it is also safe to fire on every cold open of the session entry route: the
 * server's partial unique index makes a double call resolve to the same row (no lock,
 * no debounce needed on this side either, per the route's own doc comment).
 *
 * `status` on the resolved value distinguishes 201 (generated — a new plan, empty
 * progress) from 200 (resolved — the user may be mid-session). The FLE-10 design doc is
 * explicit that flattening this is wrong: a 200 must not re-animate a progress bar or
 * fire a "session started" signal. Callers branch on `result.status`, not on inspecting
 * item states themselves.
 */
export function useTodaySession() {
  const qc = useQueryClient();
  const userId = getOrCreateUserId();

  return useMutation({
    mutationFn: (body?: { song_id?: number; target_minutes?: number }) =>
      apiFetchWithStatus<PracticeSessionResponse>('/api/v1/practice-sessions/today', {
        method: 'POST',
        body: JSON.stringify(body ?? {}),
      }),

    onSuccess: ({ data: session }) => {
      cacheSession(qc, userId, session);
    },
  });
}

/**
 * The caller's open session for today, read-only (FLE-21 §5). 404 is the normal answer
 * on a day with no session yet — not an error state, so callers should check
 * `query.isError` only to distinguish "no session" (404) from a real failure if they
 * need to; most callers just want `query.data` to decide whether to show a resume banner.
 *
 * Read-only by contract: this hook must never be the thing that starts a session. Use
 * useTodaySession for that.
 */
export function useCurrentPracticeSession() {
  const userId = getOrCreateUserId();
  const day = localCalendarDay();

  return useQuery({
    queryKey: currentPracticeSessionQueryKey(userId, day),
    queryFn: () => apiFetch<PracticeSessionResponse>('/api/v1/practice-sessions/current'),
    retry: false, // a 404 here is a normal "nothing to resume" answer, not a transient failure
  });
}

/**
 * Read back one session by id, terminal or not. Used for the summary screen (which
 * already holds the id from the completion response) and for resuming a session the
 * walker was mid-render for when the app restarted cold with a route param but no
 * warm cache.
 */
export function usePracticeSession(sessionId: string | null) {
  return useQuery({
    queryKey: practiceSessionQueryKey(sessionId ?? ''),
    queryFn: () => apiFetch<PracticeSessionResponse>(`/api/v1/practice-sessions/${sessionId}`),
    enabled: sessionId != null,
  });
}

/**
 * Mark the session opened. Write-once, idempotent server-side (FLE-21 R5) — safe to
 * call on every resume, not just the first open; a retry must not move `started_at`.
 */
export function useStartPracticeSession(sessionId: string) {
  const qc = useQueryClient();
  const userId = getOrCreateUserId();

  return useMutation({
    mutationFn: () =>
      apiFetch<PracticeSessionResponse>(`/api/v1/practice-sessions/${sessionId}/start`, {
        method: 'POST',
      }),
    onSuccess: (session) => cacheSession(qc, userId, session),
  });
}

/**
 * Item -> in_progress. Fired when an item becomes the active one — not on mount of the
 * walker route, but on the transition into that item, so a late-arriving `enter` from
 * an offline replay can't resurrect an item the user already finished
 * (`applied: false` on that path — callers should treat it as a no-op, not an error).
 */
export function useEnterItem(sessionId: string) {
  return useMutation({
    mutationFn: (itemIndex: number) =>
      apiFetch<ItemEventResponse>(
        `/api/v1/practice-sessions/${sessionId}/items/${itemIndex}/enter`,
        { method: 'POST' },
      ),
  });
}

interface CompleteItemBody {
  itemIndex: number;
  rating: PracticeRatingLiteral | null;
  active_seconds: number;
  completed_reps?: number | null;
  advance_mode?: 'user_tap' | 'auto' | null;
}

/**
 * Item -> completed, with or without a rating (`rating: null` is valid and expected —
 * warm-up, song_play, consolidation, and any slot the MAX_RATING_TAPS cap dropped are
 * unrated by design). This is the player's ONLY rating write; the server fans out to
 * drill_attempts / the user_sessions daily verdict, so callers must not also call
 * useSubmitDrillRating or useSubmitRating from sessions.ts for a session item.
 *
 * A non-null rating on the rated repertoire item can 409 if the daily verdict was
 * already recorded (e.g. the user also rated the song from the Today card). Per
 * UI-SPEC §10's existing meaning that is a state-sync signal, not an error — the
 * item's rating is already committed when the 409 is thrown, so this treats 409 as a
 * normal ItemEventResponse rather than rejecting the mutation.
 */
export function useCompleteItem(sessionId: string) {
  const qc = useQueryClient();
  const userId = getOrCreateUserId();

  return useMutation({
    mutationFn: async ({ itemIndex, ...body }: CompleteItemBody) => {
      try {
        return await apiFetch<ItemEventResponse>(
          `/api/v1/practice-sessions/${sessionId}/items/${itemIndex}/complete`,
          { method: 'POST', body: JSON.stringify(body) },
        );
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        if (msg.includes('HTTP 409')) {
          // The write committed; the 409 body is still an ItemEventResponse (schema.d.ts).
          // apiFetch discards non-2xx bodies, so re-fetch the session to resync state
          // rather than parse the thrown error's message for JSON.
          const session = await apiFetch<PracticeSessionResponse>(
            `/api/v1/practice-sessions/${sessionId}`,
          );
          cacheSession(qc, userId, session);
        }
        throw err;
      }
    },
    onSuccess: (event) => {
      cacheSession(qc, userId, event.session);
    },
  });
}

interface SkipItemBody {
  itemIndex: number;
  active_seconds?: number;
  completed_reps?: number | null;
}

/**
 * Item -> skipped. Its own call, fired only on a deliberate user skip — never inferred
 * from forward navigation (FLE-21 §5.4). This is the endpoint the pilot's "which block
 * do they skip" readout is built from; a walker that calls this on ordinary advance
 * would turn that number into noise.
 */
export function useSkipItem(sessionId: string) {
  const qc = useQueryClient();
  const userId = getOrCreateUserId();

  return useMutation({
    mutationFn: ({ itemIndex, ...body }: SkipItemBody) =>
      apiFetch<ItemEventResponse>(
        `/api/v1/practice-sessions/${sessionId}/items/${itemIndex}/skip`,
        { method: 'POST', body: JSON.stringify({ active_seconds: 0, ...body }) },
      ),
    onSuccess: (event) => cacheSession(qc, userId, event.session),
  });
}

/**
 * Session -> completed, with the summary numbers (FLE-10 R8). Outranks the clock
 * (FLE-21 §3.1): even a session the day-roll sweep already abandoned underneath the
 * user can still be completed here, and that call is the one that's believed.
 *
 * `completion_ratio`, `done_items`, `skipped_items`, `planned_items` come from the
 * server — the summary screen displays them, it does not recompute a ratio from item
 * states it read itself (there must be exactly one definition of this number; it feeds
 * mode rule 3).
 */
export function useCompletePracticeSession(sessionId: string) {
  const qc = useQueryClient();
  const userId = getOrCreateUserId();

  return useMutation({
    mutationFn: () =>
      apiFetch<SessionCompleteResponse>(`/api/v1/practice-sessions/${sessionId}/complete`, {
        method: 'POST',
      }),
    onSuccess: (result) => cacheSession(qc, userId, result.session),
  });
}
